### PR TITLE
SubItems and more

### PR DESCRIPTION
**Pull request by MFlisar**
Following was added:

**SubItems**
* `SubItemUtils` (count, selection, delete functions that respect collapsed sub items)
* Demo activity added that shows those new functions in action
* Adjusted other examples to fulfill the new two way relation between `IExpandable` and `ISubItem`

**Others**
* `RangeSelectorHelper` - it tracks long pressed items and allows to select all items between the two consecutive long pressed items
* added ability to reset and recreate the map of all possible types for the `RecyclerView` (necessary, if don't want to change your items but you have items that are able to render different item types depending on a user selected theme for example)
* `ISelectionListener` - this allows to be notified about selection changes (regardless of where they come from!)

### CODE REVIEW

**Comment by MFlisar** [comment]
In the meantime, I understand the handling of sub items, but it would be a good idea to add this to the readme. It's not clear, that sub items are temp. items and that they are not fully supported.

Additionally, to get full working sub items and a full working `SubItemUtil`, it's necessary to use the id based state management (would be good for the readme as well)

Let me know if you have questions or want some changes...

**Comment by mikepenz** [comment]
@MFlisar could you please fix the build: https://travis-ci.org/mikepenz/FastAdapter/builds/161855519#L1848

**Comment by FabianTerhorst** [comment]
Why you removed the chaining after object initialization on some places?

**Comment by MFlisar** [comment]
@FabianTerhorst 

the problem is the full generic `AbstractItem` interface. If inherited classes don't override functions (like `withIdentifier` for example), the chaining returns an `Item extends AbstractItem<?, ?>` instead of the real class which breaks chaining - you can't call functions from the real item anymore...

Either you cast after such calls or you override the functions and just call super or you break the chain...

Imho, the `AbstractItem` should not be that generic, it may make sense to improve that...

@mikepenz 
I'll do that. I have just tried the normal extensions and app

File: library/src/main/java/com/mikepenz/fastadapter/ISelectionListener.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+ */
+
+public interface ISelectionListener {
+    /**
+     * is called, whenever the provided item is selected or deselected
+     *
+     * @param item the item who's selection state changed
+     * param selected the new selection state of the item
+     */
+    void onSelectionChanged(IItem item, boolean selected);
```
Comment (mikepenz) [inline library/src/main/java/com/mikepenz/fastadapter/ISelectionListener.java]: Shouldn't this also have a generic type?

Comment (MFlisar) [inline library/src/main/java/com/mikepenz/fastadapter/ISelectionListener.java]: You will have adapters with multiple item types, but one listener. So I think, IItem should be fine. What would you suggest?

Comment (mikepenz) [inline library/src/main/java/com/mikepenz/fastadapter/ISelectionListener.java]: If an adapter is typed you will just have this specific type in general. You still can have it non-typed which will then result in `IItem` (that's the same for the adapters)

**Comment by mikepenz** [comment]
By the way you are from austria?

**Comment by MFlisar** [comment]
Yes, I am

**Comment by mikepenz** [comment]
@MFlisar can you contact me via hangouts pls?

File: library/src/main/java/com/mikepenz/fastadapter/ISelectionListener.java (hunk 1 above)
Comment (MFlisar) [inline library/src/main/java/com/mikepenz/fastadapter/ISelectionListener.java]: Then I have to change AdapterUtil.getAllItems() to be typed as well... But sure, I can do that.

**Comment by mikepenz** [comment]
The main problem why the chaining for the `SampleItem` is no longer working is that it is used for many different use cases in the sample. Which required it to be `SampleItem<S extends IItem & IExpandable>` as a result of this java can no longer resolve the proper types. Either we create more specific items for the sample, or we make the `SampleItem` abstract and add empty implementations of it with the proper type speficications.

**Comment by mikepenz** [comment]
I will merge this now and will make some improvements. so we can built upon this. It will be quite hard if we continue with this PR as it is already very big

**Pull request merged**